subsys/bluetooth/shell/has_client.c

```diff
@@ -0,0 +1,158 @@
+#include <bluetooth/audio/has.h>
+#include <shell/shell.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "bt.h"
+
+static struct bt_has *inst;
+
+static void has_client_discover_cb(struct bt_conn *conn, int err, struct bt_has *has,
+				   enum bt_has_hearing_aid_type type,
+				   enum bt_has_capabilities caps)
+{
+	if (err) {
+		shell_error(ctx_shell, "HAS discovery (err %d)", err);
+	} else {
+		shell_print(ctx_shell, "HAS discovered %p type 0x%02x caps 0x%02x for conn %p",
+			    has, type, caps, conn);
+	}
+
+	inst = has;
+}
+
+static void has_client_preset_switch_cb(struct bt_has *has, uint8_t index)
+{
+	shell_print(ctx_shell, "HAS %p preset switch index 0x%02x", has, index);
+}
+
+static void has_client_status_cb(struct bt_has *has, int err, void *user_data)
+{
+	shell_print(ctx_shell, "HAS %p status err %d user_data %p", has, err, user_data);
+}
+
+static void has_client_preset_event_cb(struct bt_has *has, enum bt_has_preset_event_type type,
+				       uint8_t index, const void *data, bool is_last)
+{
+	shell_print(ctx_shell, "HAS %p preset index 0x%02x info_type %d is_last %d",
+		    has, index, type, is_last);
+
+	if (type == BT_HAS_PRESET_RECORD_UPDATE || type == BT_HAS_PRESET_RECORD_READ_RSP) {
+		const struct bt_has_preset_record_val *value = data;
+
+		shell_print(ctx_shell, "\trecord: properties 0x%02x name %s", value->properties,
+			    value->name);
+	}
+
+	if (type == BT_HAS_PRESET_DELETE) {
+		const struct bt_has_preset_delete_val *value = data;
+
+		shell_print(ctx_shell, "\t delete: start_index 0x%02x end_index 0x%02x",
+			    index, value->end_index);
+	}
+}
+
+static const struct bt_has_client_cb has_client_cb = {
+	.discover = has_client_discover_cb,
+	.preset_switch = has_client_preset_switch_cb,
+	.status = has_client_status_cb,
+	.preset_event = has_client_preset_event_cb,
+};
+
+static int cmd_has_client_init(const struct shell *sh, size_t argc, char **argv)
+{
+	int err;
+
+	if (!ctx_shell) {
+		ctx_shell = sh;
+	}
+
+	err = bt_has_client_cb_register(&has_client_cb);
+	if (err != 0) {
+		shell_print(sh, "bt_has_client_cb_register (err %d)", err);
+	}
+
+	return err;
+}
+
+static int cmd_has_client_discover(const struct shell *sh, size_t argc, char **argv)
+{
+	int err;
+
+	if (default_conn == NULL) {
+		shell_error(sh, "Not connected");
+		return -ENOEXEC;
+	}
+
+	if (!ctx_shell) {
+		ctx_shell = sh;
+	}
+
+	err = bt_has_client_discover(default_conn);
+	if (err != 0) {
+		shell_print(sh, "bt_has_client_discover (err %d)", err);
+	}
+
+	return err;
+}
+
+static int cmd_has_client_read_presets(const struct shell *sh, size_t argc, char **argv)
+{
+	int err;
+	const uint8_t index = shell_strtoul(argv[1], 16, &err);
+	const uint8_t count = shell_strtoul(argv[2], 10, &err);
+
+	if (err < 0) {
+		shell_print(sh, "Invalid command parameter (err %d)", err);
+		return err;
+	}
+
+	if (default_conn == NULL) {
+		shell_error(sh, "Not connected");
+		return -ENOEXEC;
+	}
+
+	if (!ctx_shell) {
+		ctx_shell = sh;
+	}
```

Comment:
Suggest to replace this with `inst` - i.e. verify that `inst` is not NULL. If `inst` is not NULL, that means that discovery has completed and that the `ctx_shell` is also set.